Cut Regular Grid

santis19 opened the pull request:
gridder.cut_regular function added.

When we have a regular grid and we want to cut a rectangular area it may be usefull to have the shape of the new cutted area. This is what this function does. Entirely based on the original gridder.cut function with some additional lines.

Also, it contains the correction to the gridder.cut function that I pulled recently.

## Checklist:

- [ ] Make tests for new code
- [ ] Create/update docstrings
- [ ] Include relevant equations and citations in docstrings
- [x] Code follows PEP8 style conventions
- [ ] Code and docs have been spellchecked
- [ ] Include new dependencies in docs, requirements.txt, README, and .travis.yml
- [ ] Documentation builds properly
- [ ] All tests pass
- [ ] Can be merged
- [ ] Changelog entry (leave for last)
- [ ] First-time contributor? Add yourself to `doc/contributors.rst` (leave for last)

leouieda commented on fatiando/gridder.py:
```diff
@@ -394,3 +395,55 @@ def cut(x, y, scalars, area):
+    if xmin > xmax:
+        xmin, xmax = xmax, xmin
+    if ymin > ymax:
+        ymin, ymax = ymax, ymin
```
@santis19 the comment for these 4 lines is the same as for #188 (use `assert` to raise an error)

leouieda commented on fatiando/gridder.py:
```diff
@@ -394,3 +395,55 @@ def cut(x, y, scalars, area):
+    The returned subsection is not a copy! In technical terms, returns a slice
+    of the numpy arrays. So changes made to the subsection reflect on the
+    original grid. Use numpy.copy to make copies of the subsections and avoid
+    this.
```
This is a great docstring!

Could you a `copy` optional argument to the function so that `copy==True` returns a copy and `copy==False` returns a slice? This might be useful if you need to process the slice but don't want to mess with the original data. Putting the option in the function means writing the `numpy.copy` code only once.

leouieda commented on fatiando/gridder.py:
```diff
@@ -394,3 +395,55 @@ def cut(x, y, scalars, area):
+    * ``[subx, suby, subscalars,shape_window]``
+        Arrays with x and y coordinates and scalar values of the subsection.
+    shape_window: tuple(ny,nx)
+        Shape of the cutted grid.
```
Maybe "subgrid" would be better than "cutted grid".

leouieda commented on fatiando/gridder.py:
```diff
@@ -394,3 +395,55 @@ def cut(x, y, scalars, area):
+        xmin, xmax = xmax, xmin
+    if ymin > ymax:
+        ymin, ymax = ymax, ymin
+
```
There is another sanity check that needs to made for the input arguments. All input arrays need to be able to be broadcast into the specified `shape` argument. A quick way to check this is to do something like `assert x.size == shape[0]*shape[1]`.

leouieda commented on fatiando/gridder.py:
```diff
@@ -394,3 +395,55 @@ def cut(x, y, scalars, area):
+
+    # Calculates the distance between neighbour points
+    if shape[0]*shape[1] != len(x):
+        raise ValueError("The shape isn't the shape of the x,y grid.")
```
Noticed this line after I wrote the comment above. This should be substituted for the `assert`s described above.

leouieda commented on fatiando/gridder.py:
```diff
@@ -394,3 +395,55 @@ def cut(x, y, scalars, area):
+        raise ValueError("x and y must have the same length")
+    inside = [i for i in xrange(len(x))
+              if x[i] >= xmin and x[i] <= xmax
+              and y[i] >= ymin and y[i] <= ymax]
```
A nice thing about dealing with regular grids is that you don't have to do these calculations. Calculating `inside` can be time consuming for large grids and is not really straight forward. This is something that has to be done for irregular data, but grid data offer a better way.

If you `reshape` the input arrays into 2D arrays (`xg = x.reshape(shape)`), then you can use numpy slicing to get the subgrid. First you'll need to calculate the max and min i, j indexes of the subgrid (using the spacing between points). Once you have imin, imax, jmin, jmax, getting the subgrid becomes `subx = xg[imin:imax, jmin:jmax]`. The shape of the subgrid is `subx.shape` and you can return 1D arrays with `subx.ravel()`.

santis19 commented:
Hi @leouieda!

I've changed the "cutted grid" lines with "subgrid", thanks for noticing it, it's better that way!

I've also incorporated the assertion structure to ensure that the x,y arrays correspond with the shape tuple and also with every scalar. It would be great to do the ensure that every x,y and every scalar in the scalars list are numpy arrays, because otherwise the return line raises an error (because the lists' indexing can be done only with integers, not lists). What is the better way to do that using the assert command?

Another thing I noticed is that the function always returns a copy (not a slice) of the x,y and scalars arrays. The x[inside] in a main code IS a copy, but when we use it inside a function it returns a copy of it. I think that's because it uses the input variables as internal variables and the returned ones have nothing to do with the inputs. Should be tested, though!

I'll get into the optimization you mentioned, but I don't know if I should fork a new branch or overwrite the current file, just not to make a mess with different codes.

Greetings!

leouieda commented:
> I've also incorporated the assertion structure to ensure that the x,y arrays correspond with the shape tuple and also with every scalar. Also, it would be great to ensure that every x,y and every scalar in the scalars list are numpy arrays, because otherwise the return line raises an error (because the lists' indexing can be done only with integers, not lists). What is the better way to do that using the assert command?

You can do that using `np.atleast_1d` on the input to make sure they will be arrays. In case they already are arrays, the function won't copy the memory.

leouieda commented:
> Another thing I noticed is that the function always returns a copy (not a slice) of the x,y and scalars arrays. The x[inside] in a main code IS NOT a copy (is a slice of it), but when we use it inside a function it returns a copy of it. I think that's because it uses the input variables as internal variables and the returned ones have nothing to do with the inputs. Should be tested, though!

Wow, this is very weird. I never noticed it. Could you post an example that fails because of this?

leouieda commented:
> I'll get into the optimization you mentioned, but I don't know if I should fork a new branch or overwrite the current file, just not to make a mess with different codes.

It's fine to push these modifications here because they are completely related. Just avoid pushing edits in unrelated code (like typo fixes in other files, etc).

leouieda commented:
I'm closing this Pull Request because it's been inactive for a long time. The branch is probably very out of sync with `master`.

If you want to continue working of this, please [sync your fork](https://help.github.com/articles/syncing-a-fork/), update your branch, and re-submit the PR.  Let me know if you need any help doing this!

Sorry for the hassle! This is not a critique of your work, just a way to keep things fresh in the project. I hope you'll consider re-submitting.

**This is a message template.**
